consumer.go

```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
+			child.sendError(err)
+			child.trigger <- none{}
+		}
 	}
 }
 
-// Validate checks a ConsumerConfig instance. It will return a
-// ConfigurationError if the specified value doesn't make sense.
-func (config *ConsumerConfig) Validate() error {
-	if config.DefaultFetchSize <= 0 {
-		return ConfigurationError("Invalid DefaultFetchSize")
+func (w *consumerWorker) fetchNewMessages() (*FetchResponse, error) {
+	request := &FetchRequest{
+		MinBytes:    w.consumer.config.MinFetchSize,
+		MaxWaitTime: int32(w.consumer.config.MaxWaitTime / time.Millisecond),
 	}
 
-	if config.MinFetchSize <= 0 {
-		return ConfigurationError("Invalid MinFetchSize")
+	for child, _ := range w.work {
+		request.AddBlock(child.topic, child.partition, child.offset, child.fetchSize)
 	}
 
-	if config.MaxMessageSize < 0 {
-		return ConfigurationError("Invalid MaxMessageSize")
+	return w.broker.Fetch(w.consumer.client.id, request)
+}
+
+func (w *consumerWorker) handleResponse(child *PartitionConsumer, block *FetchResponseBlock) {
+	switch block.Err {
+	case NoError:
+		break
+	case UnknownTopicOrPartition, NotLeaderForPartition, LeaderNotAvailable:
+		// doesn't belong to us, redispatch it
+		child.trigger <- none{}
+		delete(w.work, child)
+		return
+	default:
+		child.sendError(block.Err)
+		return
 	}
 
-	if config.MaxWaitTime < 1*time.Millisecond {
-		return ConfigurationError("Invalid MaxWaitTime, it needs to be at least 1ms")
-	} else if config.MaxWaitTime < 100*time.Millisecond {
-		Logger.Println("ConsumerConfig.MaxWaitTime is very low, which can cause high CPU and network usage. See documentation for details.")
-	} else if config.MaxWaitTime%time.Millisecond != 0 {
-		Logger.Println("ConsumerConfig.MaxWaitTime only supports millisecond precision; nanoseconds will be truncated.")
+	if len(block.MsgSet.Messages) == 0 {
+		// We got no messages. If we got a trailing one then we need to ask for more data.
+		// Otherwise we just poll again and wait for one to be produced...
+		if block.MsgSet.PartialTrailingMessage {
+			if child.config.MaxMessageSize > 0 && child.fetchSize == child.config.MaxMessageSize {
+				// we can't ask for more data, we've hit the configured limit
+				child.sendError(MessageTooLarge)
+				child.offset++ // skip this one so we can keep processing future messages
+			} else {
+				child.fetchSize *= 2
+				if child.config.MaxMessageSize > 0 && child.fetchSize > child.config.MaxMessageSize {
+					child.fetchSize = child.config.MaxMessageSize
+				}
+			}
+		}
+
+		return
 	}
 
-	if config.EventBufferSize < 0 {
-		return ConfigurationError("Invalid EventBufferSize")
+	// we got messages, reset our fetch size in case it was increased for a previous request
+	child.fetchSize = child.config.DefaultFetchSize
+
+	atLeastOne := false
+	prelude := true
+	for _, msgBlock := range block.MsgSet.Messages {
+
+		for _, msg := range msgBlock.Messages() {
+			if prelude && msg.Offset < child.offset {
+				continue
+			}
+			prelude = false
```

Comment:
If we ask for offset X, kafka is allowed to send us a few messages <X if it already has e.g. a buffer in memory from X-5 to X+5, so we have to just ignore those messages. This happens in particular when using compression, since messages get compressed in batches and kafka is just going to send us the entire batch containing whichever offset we ask for.